File: src/vehicle_controller/src/waypoint_loader_node.cpp
#include "geometry_msgs/msg/pose_array.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "vehicle_controller/waypoint_loader.hpp"

using namespace std::chrono_literals;

class WaypointLoaderNode : public rclcpp::Node
{
   public:
    WaypointLoaderNode() : Node("waypoint_loader")
    {
        // Declare parameters
        this->declare_parameter("csv_file", "");
        this->declare_parameter("publish_frequency", 10.0);  // Hz
        this->declare_parameter("frame_id", "map");

        // Get parameters
        std::string csv_file = this->get_parameter("csv_file").as_string();
        double publish_frequency = this->get_parameter("publish_frequency").as_double();
        frame_id_ = this->get_parameter("frame_id").as_string();

        // Create publishers
        waypoints_publisher_ = this->create_publisher<geometry_msgs::msg::PoseArray>("waypoints", 10);

        waypoint_names_publisher_ = this->create_publisher<std_msgs::msg::String>("waypoint_names", 10);

        // Create waypoint loader
        waypoint_loader_ = std::make_unique<WaypointLoader>(this->get_logger());

        if (!csv_file.empty())
        {
            waypoints_ = waypoint_loader_->load_waypoints(csv_file);

            if (!waypoints_.empty())
            {
                RCLCPP_INFO(this->get_logger(), "Loaded %zu waypoints from %s", waypoints_.size(), csv_file.c_str());

                // Create timer to publish waypoints
                timer_ = this->create_wall_timer(std::chrono::milliseconds(static_cast<int>(1000.0 / publish_frequency)),
                                                 std::bind(&WaypointLoaderNode::publish_waypoints, this));
            }
            else
            {
                RCLCPP_ERROR(this->get_logger(), "No waypoints loaded from %s", csv_file.c_str());
            }
        }
        else
        {
            RCLCPP_ERROR(this->get_logger(), "No CSV file specified. Use the 'csv_file' parameter.");
        }
    }

   private:
    std::unique_ptr<WaypointLoader> waypoint_loader_;
    std::vector<Waypoint> waypoints_;
    rclcpp::Publisher<geometry_msgs::msg::PoseArray>::SharedPtr waypoints_publisher_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr waypoint_names_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::string frame_id_;

    void publish_waypoints()
    {
        if (waypoints_.empty())
        {
            return;
        }

        // Create PoseArray message
        auto pose_array_msg = std::make_unique<geometry_msgs::msg::PoseArray>();
        pose_array_msg->header.stamp = this->now();
        pose_array_msg->header.frame_id = frame_id_;

        // Create names message
        auto names_msg = std::make_unique<std_msgs::msg::String>();
        std::string names_str;

        // Add all waypoints to the pose array
        for (const auto &waypoint : waypoints_)
        {
            geometry_msgs::msg::Pose pose;
            pose.position.x = waypoint.x;
            pose.position.y = waypoint.y;
            pose.position.z = 0.0;

            // Set orientation to identity quaternion (no rotation)
            pose.orientation.x = 0.0;
            pose.orientation.y = 0.0;
            pose.orientation.z = 0.0;
            pose.orientation.w = 1.0;

            pose_array_msg->poses.push_back(pose);

            // Add name to names string (comma-separated)
            if (!names_str.empty())
            {
                names_str += ",";
            }
            names_str += waypoint.name.empty() ? "unnamed" : waypoint.name;
        }

        names_msg->data = names_str;

        // Publish messages
        waypoints_publisher_->publish(std::move(pose_array_msg));
        waypoint_names_publisher_->publish(std::move(names_msg));
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<WaypointLoaderNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/vehicle_controller_py/vehicle_controller_py/waypoint_loader_node.py
import os
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseArray, Pose
from std_msgs.msg import String
from dataclasses import dataclass
from typing import List
from ament_index_python.packages import get_package_share_directory

@dataclass
class Waypoint:
    """Simple data structure to hold waypoint information."""
    x: float
    y: float
    name: str = ""

class WaypointLoaderNode(Node):
    def __init__(self):
        super().__init__('waypoint_loader')

        # 1. Declare and Get Parameters
        self.declare_parameter('csv_file', 'coordinates.csv')
        self.declare_parameter('publish_frequency', 10.0)  # Hz
        self.declare_parameter('frame_id', 'map')

        csv_filename = self.get_parameter('csv_file').value
        publish_frequency = self.get_parameter('publish_frequency').value
        self.frame_id = self.get_parameter('frame_id').value

        # Dynamically build the path to the CSV file 
        try:
            package_share_directory = get_package_share_directory('vehicle_controller_py')
            self.csv_file = os.path.join(package_share_directory, 'config', csv_filename)
        except Exception as e:
            self.get_logger().error(f"Could not find package share directory: {e}")
            self.csv_file = csv_filename # Fallback to just the filename

        # 2. Create Publishers
        self.waypoints_publisher = self.create_publisher(PoseArray, 'waypoints', 10)
        self.waypoint_names_publisher = self.create_publisher(String, 'waypoint_names', 10)

        # 3. Load Waypoints
        self.waypoints = self.load_waypoints(self.csv_file)

        # 4. Setup Timer if Waypoints Loaded Successfully
        if self.waypoints:
            self.get_logger().info(f"Loaded {len(self.waypoints)} waypoints from {self.csv_file}")
            
            # Calculate timer period in seconds
            period = 1.0 / publish_frequency if publish_frequency > 0 else 0.1
            self.timer = self.create_timer(period, self.publish_waypoints)
        else:
            if self.csv_file:
                self.get_logger().error(f"No waypoints loaded from {self.csv_file}")
            else:
                self.get_logger().error("No CSV file specified. Use the 'csv_file' parameter.")

    def load_waypoints(self, filename: str) -> List[Waypoint]:
        """Reads the coordinates.csv file and parses the X, Y, and optional name."""
        waypoints = []
        if not filename:
            return waypoints

        try:
            with open(filename, 'r') as f:
                self.get_logger().info(f"Loading waypoints from: {filename}")
                
                for line in f:
                    line = line.strip()
                    
                    # Skip comments or empty lines
                    if not line or line.startswith('#'):
                        continue

                    parts = line.split(',')
                    
                    # Ensure we have at least X and Y
                    if len(parts) >= 2:
                        try:
                            x = float(parts[0].strip()) / 1000.0
                            y = float(parts[1].strip()) / 1000.0
                            
                            # Safely extract name if it exists, otherwise leave empty
                            name = parts[2].strip() if len(parts) > 2 else ""
                            
                            waypoints.append(Waypoint(x=x, y=y, name=name))
                        except ValueError:
                            # Skip malformed lines where conversion to float fails
                            continue 
                            
        except Exception as e:
            self.get_logger().error(f"Failed to open file: {filename}. Error: {e}")

        self.get_logger().info(f"Loaded {len(waypoints)} waypoints")
        return waypoints

    def publish_waypoints(self):
        """Timer callback to publish the PoseArray and string of names."""
        if not self.waypoints:
            return

        # Prepare PoseArray message
        pose_array_msg = PoseArray()
        pose_array_msg.header.stamp = self.get_clock().now().to_msg()
        pose_array_msg.header.frame_id = self.frame_id

        names_list = []

        # Populate messages
        for wp in self.waypoints:
            pose = Pose()
            pose.position.x = wp.x
            pose.position.y = wp.y
            pose.position.z = 0.0

            # Set orientation to identity quaternion (no rotation)
            pose.orientation.x = 0.0
            pose.orientation.y = 0.0
            pose.orientation.z = 0.0
            pose.orientation.w = 1.0

            pose_array_msg.poses.append(pose)
            
            # Store name (or "unnamed" if empty)
            names_list.append(wp.name if wp.name else "unnamed")

        # Prepare String message (comma-separated names)
        names_msg = String()
        names_msg.data = ",".join(names_list)

        # Publish
        self.waypoints_publisher.publish(pose_array_msg)
        self.waypoint_names_publisher.publish(names_msg)

def main(args=None):
    rclpy.init(args=args)
    node = WaypointLoaderNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()